### PR TITLE
fix: add bounded cancellable worker response delivery retries

### PR DESCRIPTION
**Pull request by YunchuWang**
# Summary

## What changed?
- Add one shared worker delivery loop for orchestration/activity/entity completion, version-mismatch failure completion, and orchestration abandonment. Retry the same response/token, never user execution.
- Match .NET's **10 SDK attempts** for `UNAVAILABLE`, `UNKNOWN`, `DEADLINE_EXCEEDED`, and `INTERNAL`: 200 ms exponential backoff, capped at 15 seconds before +0–20% jitter. Permanent/exhausted failures retain existing error logs.
- Cancel retry waits/retried RPCs on stop while allowing initial completion during the existing graceful-drain window. Retired stubs wait only for their own pending work; original-run cancellation survives reconnect/restart.
- Make the existing unary helper's metadata wait abort-aware, with cleanup and no late RPC after cancellation.
- Route version rejection through private `_abandonOrchestrationWorkItem(stub, completionToken, signal?)`, backed by the same delivery loop. An explicit signal governs initial attempts, retries and backoff; omission preserves graceful-initial/retry-stop behavior. This provides the shared primitive for the independently based history PR, not another retry framework.
- **Preserve caller channel options and Azure-managed transport retry configuration**, matching the .NET Azure-managed worker. Ten SDK attempts is not a ten-physical-request guarantee.
- Explicitly run both real-socket and persisted-backend delivery specs in the existing DTS emulator group on **Node 22/24**. The backend suite remains opt-in locally; CI supplies its emulator connection string.

## Why is this change needed?
Azure-managed workers already configure transport retries, normally for UNAVAILABLE. Core completion/abandon sites lack a bounded SDK delivery loop, including after headers commit and for additional .NET transient statuses. This closes that gap without changing persisted history or the backend's at-least-once work-item delivery contract.

Pinned reference: `microsoft/durabletask-dotnet@bc2bc12ca5ee3a12a6e633250ade5efe6ae90ef4`, processor `ExecuteWithRetryAsync`, internal options, and `GrpcBackoff.cs`. Its Azure-managed `CreateChannel` also retains `GrpcRetryPolicyDefaults.DefaultServiceConfig` beneath that loop, confirming the layered retry behavior.

Base: upstream main `28730dfaedaecd696468cae2e2365153b230fa56`. **Final tested head: `4ce281c32f46ca806fcd39acc8f86b2a8b1e7abe`. All final-head CI workflows passed. The coordinator's principal source re-review approved this exact head, including the shared-abandonment delta; no remaining source blocker.**

## Issues / work items
- Focused worker-response parity change. No client wait feature, history hydration, chunking, or concurrency implementation.

---

# Project checklist
- [ ] Release notes are not required for the next release
  - [x] Otherwise: Notes added to `CHANGELOG.md`
- [x] Backport is not required
  - [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
- [x] All required tests have been added/updated (unit tests, E2E tests)
- [ ] Breaking change?
  - [ ] If yes:
    - Impact: No public API/wire-format changes; existing channel configuration is preserved.
    - Migration guidance: Account for both SDK and configured transport retry layers as documented in README.

---

# AI-assisted code disclosure (required)

## Was an AI tool used? (select one)
- [ ] No
- [ ] Yes, AI helped write parts of this PR (e.g., GitHub Copilot)
- [x] Yes, an AI agent generated most of this PR

If AI was used:
- Tool(s): GitHub Copilot CLI.
- AI-assisted areas/files: Implementation, tests, CI selector wiring, documentation.
- What you changed after AI output: Restricted channel retention to per-stub work; wired explicit CI coverage; removed a transport-retry override after inspecting .NET layering; fixed pending metadata cancellation; extracted shared abandonment to close a concrete cross-PR coverage gap. Principal source re-review approved the final head; human confirmation below remains separate.

AI verification (required if AI was used):
- [ ] I understand the code and can explain it (human confirmation pending)
- [x] I verified referenced APIs/types exist and are correct
- [x] I reviewed edge cases/failure paths (timeouts, retries, cancellation, exceptions)
- [x] I reviewed concurrency/async behavior
- [x] I checked for unintended breaking or behavior changes

---

# Testing

## Automated tests
- **404 passed / 19 local suites at final head**, using existing Jest: `npx --no-install jest --runInBand --runTestsByPath ... --detectOpenHandles`. Covers affected worker/versioning/entity/tracing/host-integration/backoff behavior, Azure-managed builder/retry/options/endpoint behavior, and client regressions for the shared unary helper.
- Includes **21 real-gRPC socket cases**, using OS-assigned loopback ports: committed-header transient failures, permanent failures, bounded delivery, shutdown metadata/RPC/backoff cancellation, graceful initial completion, all existing completion/abandon paths, reconnect ownership, and explicit-signal abandonment.
- Five explicit-signal tests first failed during initial metadata/RPC, backoff and retry metadata/RPC; they now pass. New socket cases prove transient abandonment reuses its payload/token and explicit cancellation stops initial metadata/RPC immediately. Default drain tests remain unchanged.
- Metadata regressions first failed with pending work stuck at one. They now assert cleanup before metadata resolves, preserved initial drain, and no late RPC.
- Configured retry-layer proof: **10 SDK calls and 50 actual server calls** under a five-attempt INTERNAL channel policy, observing transport attempt metadata and one user activity execution. Only delay duration is shortened; separate deterministic tests verify exact backoff.
- **All final-head CI workflows passed**: [Test and Build](https://github.com/microsoft/durabletask-js/actions/runs/34256188373), [DTS Emulator E2E](https://github.com/microsoft/durabletask-js/actions/runs/34256188477), [Validate Samples](https://github.com/microsoft/durabletask-js/actions/runs/34256188345).
- Final DTS delivery-group jobs on [Node 22](https://github.com/microsoft/durabletask-js/actions/runs/34256188477/job/102162486264) and [Node 24](https://github.com/microsoft/durabletask-js/actions/runs/34256188477/job/102162486351) each passed **61 tests / 6 suites, zero skipped**, explicitly including both delivery specs and real emulator persisted outcomes.
- `npm run build:core`, `npm run build:azuremanaged`, changed TS ESLint/Prettier and `git diff --check`: passed. Changelog/workflow formatting passed previously and those files are unchanged by the latest delta. README's pre-existing whole-file formatting was not rewritten.

## Manual validation (only if runtime/behavior changed)
- Windows, Node.js v24.14.0; **real HTTPS Azure DTS Consumption, westus2, dedicated `workerresponse` task hub**, Azure CLI authentication. No Azure resources created/deleted; no combined-run hub access.
- Final-head command: `npx --no-install jest --runInBand --runTestsByPath test/e2e-azuremanaged/worker-response-delivery.spec.ts --detectOpenHandles`, with the dedicated `WORKER_DELIVERY_CONNECTION_STRING`. **2 passed**.

### Real Azure baseline — no injection
- Orchestration: `response-delivery-2ba5d9c0-a76e-49e1-905d-1985050d0a41`.
- Entity: `@deliverycounter-response-delivery-2ba5d9c0-a76e-49e1-905d-1985050d0a41@counter`.
- Persisted `Completed`, output `42`, entity state `42`; activity/entity each executed once.
- Four completion payloads (orchestrator x2, activity x1, entity x1), **one SDK attempt each**.

### Real Azure persisted outcome with controlled client-side faults
- Orchestration: `response-delivery-1b138691-3b2a-40aa-893e-4b2db2ba5c41`.
- Entity: `@deliverycounter-response-delivery-1b138691-3b2a-40aa-893e-4b2db2ba5c41@counter`.
- Persisted `Completed`, output `42`, entity state `42`; activity/entity each executed once.
- Four pre-send INTERNAL injections, one per completion payload; **two SDK attempts per payload**. Serialized payloads/completion tokens asserted identical across retries.
- Injection is a **test client interceptor before forwarding to real Azure**, not an actual Azure outage. SDK invocation counts are not physical network-attempt counts. Actual server-side faults and explicit-signal abandonment are covered separately by the socket suite.

---

# Notes for reviewers
- **Companion history-PR merge resolution:** both independently based PRs define `private async _abandonOrchestrationWorkItem(stub, completionToken, signal?): Promise<void>`. Keep **this PR's retry-backed helper body** and **both history-PR callers**. The history-fetch-error caller supplies its immediate-stop signal; version rejection omits it. Do not retain the history PR's standalone direct-`callWithMetadata` helper body in the combined tree, or that path bypasses response retries. Combined history-error + transient-abandon fault coverage is owned by integration; this standalone PR does not add history fetching.
- The SDK bound is an attempt bound, not a new per-RPC wall-clock deadline. Existing transport retries may multiply underlying attempts.
- Initial completion stays graceful by default; explicitly supplied abandonment cancellation covers every phase immediately. Metadata generation itself has no cancellation argument and may continue, but cannot cause a late RPC.
- Backend redelivery can still rerun user code; exactly-once assertions concern retries of a single response delivery only.
- Commit-hook shell execution hit unavailable WSL Bash; the identical `npx --no-install lint-staged` check passed when run directly before committing.

### CODE REVIEW

File: packages/durabletask-js/test/worker-response-delivery.spec.ts [hunk 1]
```diff
@@ -0,0 +1,404 @@
...
+    details: "injected delivery failure",
+    metadata: new grpc.Metadata(),
+  });
+}
+
+function unaryCall(cancel = jest.fn()): grpc.ClientUnaryCall {
+  return Object.assign(new EventEmitter(), { cancel, getPeer: () => "test", getAuthContext: () => null });
+}
+
+type CompleteCallback = (error: grpc.ServiceError | null, response: pb.CompleteTaskResponse) => void;
```
[the remark this answers is not in the document]
Comment (YunchuWang) [reply packages/durabletask-js/test/worker-response-delivery.spec.ts]: Fixed in 1e60d494fb733b4055230960cf00347eb1c868ca. CompleteCallback keeps CompleteTaskResponse as its default, while the shared completion/abandon mock accepts the response union and constructs AbandonOrchestrationTaskResponse for the abandon case. Only this test file changed; all 40 tests in the affected spec pass. Production is unchanged from 4ce281c.
